File: src/main/java/de/byjoker/myjfql/database/SimpleDatabase.java
package de.byjoker.myjfql.database;

import de.byjoker.myjfql.exception.TableException;
import de.byjoker.myjfql.util.IDGenerator;
import org.jetbrains.annotations.NotNull;

import java.io.File;
import java.util.Collection;
import java.util.HashMap;
import java.util.Map;
import java.util.Objects;

public class SimpleDatabase implements Database {

    private final String name;
    private final Map<String, Table> tables;
    private String id;
    private DatabaseType type;

    public SimpleDatabase(String name) {
        this.id = IDGenerator.generateMixed(16);
        this.name = name;
        this.tables = new HashMap<>();
        this.type = DatabaseType.SPLIT;
    }

    public SimpleDatabase(String id, String name, DatabaseType type) {
        this.id = id;
        this.name = name;
        this.tables = new HashMap<>();
        this.type = type;
    }

    @Override
    public void regenerateId() {
        this.id = IDGenerator.generateMixed(16);
    }

    @Override
    public void createTable(Table table) {
        if (getTable(table.getName()) != null)
            throw new TableException("Table already exists in database!");

        saveTable(table);
    }

    @Override
    public void saveTable(@NotNull Table table) {
        tables.put(table.getName(), table);
    }

    @Override
    public boolean existsTable(@NotNull String name) {
        return tables.containsKey(name);
    }

    @Override
    public void deleteTable(@NotNull String name) {
        if (type == DatabaseType.SPLIT) {
            new File("database/" + id + "/" + name).delete();
        }

        tables.remove(name);
    }

    @Override
    public void reformat(@NotNull DatabaseType type, DatabaseService service) {
        service.deleteDatabase(id);
        this.type = type;
        service.createDatabase(this);
    }

    @Override
    public Table getTable(@NotNull String name) {
        return tables.get(name);
    }

    @NotNull
    @Override
    public Collection<Table> getTables() {
        return tables.values();
    }

    @NotNull
    @Override
    public DatabaseType getType() {
        return type;
    }

    @NotNull
    @Override
    public String getId() {
        return id;
    }

    @NotNull
    @Override
    public String getName() {
        return name;
    }

    @Override
    public String toString() {
        return "SimpleDatabase{" +
                "name='" + name + '\'' +
                ", tables=" + tables +
                ", type=" + type +
                ", id='" + id + '\'' +
                '}';
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        SimpleDatabase that = (SimpleDatabase) o;
        return Objects.equals(id, that.id);
    }

    @Override
    public int hashCode() {
        return Objects.hash(id);
    }
}

File: src/main/java/de/byjoker/myjfql/database/SimpleDatabase.kt
package de.byjoker.myjfql.database

import de.byjoker.myjfql.exception.DatabaseException
import de.byjoker.myjfql.util.IDGenerator
import java.io.File

class SimpleDatabase(
    private var id: String = IDGenerator.generateString(12),
    private var name: String,
    private var type: DatabaseType,
) : Database {

    private val tables: MutableMap<String, Table> = mutableMapOf()

    override fun createTable(table: Table) {
        if (getTable(table.getId()) != null || getTableByName(table.getName()) != null) {
            throw DatabaseException("Table already exist!")
        }

        saveTable(table)
    }

    override fun saveTable(table: Table) {
        tables[table.getId()] = table
    }

    override fun getTable(tableId: String): Table? {
        return tables[tableId]
    }

    override fun getTableByName(name: String): Table? {
        return tables.values.firstOrNull { table -> table.getName() == name }
    }

    override fun getTableByIdentifier(identifier: String): Table? {
        if (identifier.startsWith("#")) {
            return getTable(identifier.replaceFirst("#", ""))
        }

        return getTableByName(identifier)
    }

    override fun deleteTable(tableId: String) {
        if (type == DatabaseType.INTERNAL) {
            throw DatabaseException("Internal tables can't be deleted manually!")
        }

        if (type == DatabaseType.DOCUMENT || type == DatabaseType.SHARDED) {
            File("databases/${this.id}/$tableId").delete()
        }

        tables.remove(tableId)
    }

    override fun getTables(): List<Table> {
        return tables.values.toList()
    }

    override fun format(type: DatabaseType, databaseService: DatabaseService) {
        val original = this.type

        try {
            databaseService.deleteDatabase(id)
            this.type = type
            databaseService.saveDatabase(this)
        } catch (ex: Exception) {
            databaseService.deleteDatabase(id)
            this.type = original
            databaseService.saveDatabase(this)
        }
    }

    override fun setName(name: String) {
        this.name = name
    }

    override fun getType(): DatabaseType {
        return type
    }

    override fun getName(): String {
        return name
    }

    override fun getId(): String {
        return id
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as SimpleDatabase

        if (id != other.id) return false

        return true
    }

    override fun hashCode(): Int {
        return id.hashCode()
    }

    override fun toString(): String {
        return "SimpleDatabase(id='$id', name='$name', type=$type, tables=${tables.keys}})"
    }


}
